Adding doc page for Storage.copy API (#3309)
aws-amplify/docs: docs/lib/storage/fragments/js/copy.md

`Storage.copy` method copies an existing file to a different key and returns a `{key: new S3 Object key}` object upon
success.

<amplify-callout>

`Storage.copy` can copy an object up to 5 GB in a single operation.

</amplify-callout>

To copy a file, you need to provide a source key `src` and a destination key `dest`

```typescript
await Storage.copy(
	src: { 
		key: string,
		level?: 'private' | 'protected' | 'public',

> 		identityId?: string, // only works if you are copying from a protected file 

Reviewer: ```suggestion
		identityId?: string, // only required if you are copying from a protected file 
```

	}
)
```

In addition, you can configure advanced capabilities such as cache control, meta data, and more.

```typescript
await Storage.copy({ key: 'src' }, { key: 'dest' }, {
	acl?: string, // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property
	cacheControl?: string, // Specifies caching behavior along the request/reply chain
	expires?: Date, // The date at which the object is no longer cacheable
	metadata?: [key: string]: string, // A map of metadata to store with the object in S3
})
```

## Copy files within the same access levels

You can copy a file from the specified key to another key within the same [File Access Level](~/lib/storage/configureaccess.md) (Defaults to 'public').

```typescript
console.log(await Storage.list('copied/')); // []

// Copies 'existing/srcKey' to 'copied/destKey' within the 'public' access level
const copied = await Storage.copy({ key: 'existing/srcKey' }, { key: 'copied/destKey' });

// There should now be a new file with key "copied/destKey"
console.log(await Storage.list('copied/')); // [ { ..., key: 'copied/destKey' } ]
console.log(copied); // { key: 'copied/destKey' }
```

> ## Copy file across access levels

Reviewer: ```suggestion
## Copy files across access levels
```

To copy file across different access levels, you'll need to explicitly provide the source and destination access levels:

```typescript
console.log(await Storage.list('copied/', { level: 'private' })); // []

// Copies 'existing/srcKey' to 'copied/destKey' from 'public' to 'private'
const copied = await Storage.copy(
	{ key: 'existing/srcKey', level: 'public' }, 
	{ key: 'copied/destKey', level: 'private' }
);

// There should now be a new file with key "copied/destKey" in the 'private' level
console.log(await Storage.list('copied/', { level: 'private' })); // [ { ..., key: 'copied/destKey' } ]
console.log(copied); // { key: 'copied/destKey' }
```

> ## Copy protected file from another user

Reviewer: ```suggestion
## Copy protected files from another user
```

You can also copy a protected file from another user by providing their identity id

```typescript
console.log(await Storage.list('copied/', { level: 'private' })); // []

// Copies 'existing/srcKey' to 'copied/destKey' from 'protected' of another identity ID to 'private' of the current
// authenticated user

const copied = await Storage.copy(
	{ key: 'existing/srcKey', level: 'protected', identityId: 'identityId' },
	{ key: 'copied/destKey', level: 'private' },
)

// There should now be a new file with key "copied/destKey"
console.log(await Storage.list('copied/', { level: 'private' })); // [..., key: 'copied/destKey']
console.log(copied); // { key: 'copied/destKey' }
```

> <amplify-callout>

Reviewer: ```suggestion
<amplify-callout>

```

> Cross identity ID copying is only allowed if the source object's access level is set to 'protected'.

Reviewer: ```suggestion
Cross identity ID copying is only allowed if the source object's access level is set to 'protected'.

```

> ## Copy

Reviewer: remove H2 because page header is already H2. this would render as: 
"Copy files"
"Copy"

which would look weird.

> `Storage.copy` method copies an existing file to a different key.

Reviewer: ```suggestion
`Storage.copy` method copies an existing file to a different key and returns a `{key: new S3 Object key}` object upon success.
```

> It returns a `{key: new S3 Object key}` object on success:

Reviewer: remove this line

> <amplify-callout>

Reviewer: I think in order of this to work, it needs a blank line between after `<amplify-callout>` and before `</amplify-callout>`

> Storage.copy can copy an object up to 5 GB in a single operation.

Reviewer: ```suggestion
`Storage.copy` can copy an object up to 5 GB in a single operation.
```

> </amplify-callout

Reviewer: ```suggestion
</amplify-callout>
```

> ### Options

Reviewer: Use H2

Reviewer: or actually... maybe remove this line and instead say:

"To copy a file, you need to provide a source file `src` and a destination key `dest`"

```typescript
await Storage.copy(
	src: { 
		key: string,
		level?: 'private' | 'protected' | 'public',
		identityId?: string, // only works if you are copying from a protected file 
	},
	dest: { 
		key: string,
		level?: 'private' | 'protected' | 'public',
	}, 
	config?: {
		track?: boolean, // if true, automatically send Storage Events to Amazon Pinpoint
	}
)
```

> ### Other Options

Reviewer: remove this and say: 
"In addition, you can configure advanced capabilities such as cache control, meta data, and more."

```typescript
await Storage.copy({ key: 'src' }, { key: 'dest' }, {
	acl?: string, // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property
	cacheControl?: string, // Specifies caching behavior along the request/reply chain
	expires?: Date, // The date at which the object is no longer cacheable
	metadata?: [key: string]: string, // A map of metadata to store with the object in S3
})
```

> ### Copy file

Reviewer: Use H2 and change to "Copy files within the same access levels"

> For the most basic use case, you can copy a file from the specified key to another key within the same [File Access

Reviewer: ```suggestion
You can copy a file from the specified key to another key within the same [File Access
```

> Level](~/lib/storage/configureaccess.md) (Defaults to 'public').

Reviewer: sift up to the line above. I don't think our docs will interpret this correctly.

```typescript
console.log(await Storage.list('copied/')); // []

// Copies 'existing/srcKey' to 'copied/destKey' within the 'public' access level
const copied = await Storage.copy({ key: 'existing/srcKey' }, { key: 'copied/destKey' });

// There should now be a new file with key "copied/destKey"
console.log(await Storage.list('copied/')); // [ { ..., key: 'copied/destKey' } ]
console.log(copied); // { key: 'copied/destKey' }
```

> ### Copy file across access levels

Reviewer: Use H2

> Another use case of `copy` is to copy file across different access levels

Reviewer: ```suggestion
To copy file across different access levels, you'll need to explicitly provide the source and destination access levels:
```

Reviewer: Is my interpretation above correct?

Author: Yes this is correct

```typescript
console.log(await Storage.list('copied/', { level: 'private' })); // []

// Copies 'existing/srcKey' to 'copied/destKey' from 'public' to 'private'
const copied = await Storage.copy(
	{ key: 'existing/srcKey', level: 'public' }, 
	{ key: 'copied/destKey', level: 'private' }
);

// There should now be a new file with key "copied/destKey" in the 'private' level
console.log(await Storage.list('copied/', { level: 'private' })); // [ { ..., key: 'copied/destKey' } ]
console.log(copied); // { key: 'copied/destKey' }
```

> ### Copy protected file from another identityId

Reviewer: Use H2

Reviewer: Is it better to say "user" instead of "identityId"?

Author: Agree, but I think its still worthwhile to mention identity id. I changed the line below to 'You can also copy a protected file from another user by providing their identity id', instead, let me know what you think

You can also copy a protected file from another identityId.

```typescript
console.log(await Storage.list('copied/', { level: 'private' })); // []

// Copies 'existing/srcKey' to 'copied/destKey' from 'protected' of another identity ID to 'private' of the current
// authenticated user

const copied = await Storage.copy(
	{ key: 'existing/srcKey', level: 'protected', identityId: 'identityId' },
	{ key: 'copied/destKey', level: 'private' },
)

// There should now be a new file with key "copied/destKey"
console.log(await Storage.list('copied/', { level: 'private' })); // [..., key: 'copied/destKey']
console.log(copied); // { key: 'copied/destKey' }
```

<amplify-callout>
Cross identity ID copying is only allowed if the source object's access level is set to 'protected'.
</amplify-callout>

> ### Encrypted copy

Reviewer: ```suggestion
## How to enable encrypted file copy
```

> To utilize Server-Side Encryption with AWS KMS, the following options can be passed in with the Copy API like so:

Reviewer: ```suggestion
To utilize Server-Side Encryption with AWS KMS, the following options can be passed in to `Storage.copy`:
```

`Storage.copy` method copies an existing file to a different key and returns a `{key: new S3 Object key}` object upon
success.

<amplify-callout>

`Storage.copy` can copy an object up to 5 GB in a single operation.

</amplify-callout>

To copy a file, you need to provide a source key `src` and a destination key `dest`

```typescript
await Storage.copy(
	src: { 
		key: string,
		level?: 'private' | 'protected' | 'public',

> 		identityId?: string, // only required if you are copying from a protected file 

Reviewer: ```suggestion
		identityId?: string, // only required if you are copying a protected file from another user 
```

## Copy

`Storage.copy` method copies an existing file to a different key.

It returns a `{key: new S3 Object key}` object on success:

<amplify-callout>
Storage.copy can copy an object up to 5 GB in a single operation.
</amplify-callout

### Options

```typescript
await Storage.copy(
	src: { 
		key: string;
		level?: 'private' | 'protected' | 'public';
		identityId?: string; // only works if you are copying from a protected file 
	},
	dest: { 
		key: string;
		level?: 'private' | 'protected' | 'public' 
	}, 
	config?: {
		track?: boolean, // if true, automatically send Storage Events to Amazon Pinpoint
		progressCallback?: (progress) => any, // callback function on every successful part upload
	}
)
```

> ### Extra Options

Reviewer 2: To match the wording we use in other places 

```suggestion
### Other Options
```

## Copy

`Storage.copy` method copies an existing file to a different key.

It returns a `{key: new S3 Object key}` object on success:

### Options

```typescript
await Storage.copy(
	src: { 
		key: string;
		level?: 'private' | 'protected' | 'public';
		identityId?: string; // only works if you are copying from a protected file 
	},
	dest: { 
		key: string;
		level?: 'private' | 'protected' | 'public' 
	}, 
	config?: {
		track?: boolean, // if true, automatically send Storage Events to Amazon Pinpoint
		progressCallback?: (progress) => any, // callback function on every successful part upload
	}
)
```

### Extra Options
```typescript
await Storage.copy({ key: 'src' }, { key: 'dest' }, {
	acl?: string, // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property
	cacheControl?: string, // Specifies caching behavior along the request/reply chain
	contentDisposition?: string, // Specifies presentational information for the object 
	contentEncoding?: string, // Specifies what content encodings have been applied to the object and thus what decoding mechanisms must be applied  
	contentLanguage?: string, // The language the content is in

> 	contentType?: string, // A standard MIME type describing the format of the object data

Reviewer 3: I would try to reuse it from the original object and added in case customer requires to change it
